### PR TITLE
Fix approximateOr of (A & Double) | Null

### PR DESCRIPTION
**Pull request by dwijnand**
[test_non_bootstrapped]

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 1]
```diff
@@ -478,6 +478,11 @@ object Types {
       case _ => f(this)
     }
 
+    final def mapReduceAndOr[T](f: Type => T)(g: (T, T) => T)(using Context): T = stripTypeVar match {
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/Types.scala:481]: That's logically wrong. The base classes of A & B are the union of the base classes of A and the base classes of B not the intersection. 

Is the problem is also there if the union is with a type other than `Null`?

Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/Types.scala:481]: Then I believe the fix should also be null-specific. Maybe reassign to @noti0na1 ?

Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/Types.scala:481]: I will take a look today

File: compiler/src/dotty/tools/dotc/core/TypeOps.scala [hunk 2]
```diff
@@ -225,16 +225,18 @@ object TypeOps:
...
-        cs2.foreach(cs2AsSet += _)
-        cs1.filter(cs2AsSet.contains)
+      val cs2AsSet = BaseClassSet(cs2)
+      cs1.filter(cs2AsSet.contains)
+
+    /** a version of Type#baseClasses that treats bottom types correctly */
+    def orBaseClasses(tp: Type): List[ClassSymbol] = tp.stripTypeVar match
+      case OrType(tp1, tp2) =>
+        if tp1.isBottomType && (tp1 frozen_<:< tp2) then orBaseClasses(tp2)
+        else if tp2.isBottomType && (tp2 frozen_<:< tp1) then orBaseClasses(tp1)
```
Comment (dwijnand) [inline compiler/src/dotty/tools/dotc/core/TypeOps.scala:237]: Similar to [`e65305d` (#11979)](https://github.com/lampepfl/dotty/pull/11979/commits/e65305dd2268ec245b246f90265d6db825820f0c)

File: tests/pos/i16236.scala [hunk 3]
```diff
@@ -0,0 +1,9 @@
+trait A
```
Comment (noti0na1) [inline tests/pos/i16236.scala:1]: Could you make a copy of this test and copy to `tests/explicit-nulls/pos` as well? Thanks

**Review by noti0na1 — APPROVED** [review]
LGTM, thanks!

**Comment by michelou** [comment]
@dwijnand Nice cleanup !

**Review by odersky — APPROVED** [review]
Yes, nice cleanup!

**Pull request merged**